dissect/volume/ddf/ddf.py

```diff
@@ -0,0 +1,389 @@
+        self.block_size = self.header.Block_Size
+        self.rotate_parity_count = self.header.Rotate_Parity_Count
+
+        self.physical_disk_sequence = c_ddf.uint32[num_entries](fh)[: self.primary_element_count]
+        self.starting_block = c_ddf.uint64[num_entries](fh)[: self.primary_element_count]
+
+    def __repr__(self) -> str:
+        return f"<VirtualDiskConfigurationRecord guid={self.guid}>"
+
+
+def _read_section_header(fh: BinaryIO, structure: Structure, signature: int) -> Instance:
+    obj = structure(fh)
+    if obj.Signature != signature:
+        raise DDFError(f"Invalid {structure.name} header. Signature: {obj.Signature:#010x}, expected {signature:#x}")
+    return obj
+
+
+def _read_physical_disk_records(fh: BinaryIO, offset: int) -> list[PhysicalDiskRecord]:
+    fh.seek(offset)
+    header = _read_section_header(fh, c_ddf.Physical_Disk_Records, c_ddf.DDF_PDR_SIGNATURE)
+    return [PhysicalDiskRecord(fh) for _ in range(header.Populated_PDEs)]
+
+
+def _read_virtual_disk_records(fh: BinaryIO, offset: int) -> list[VirtualDiskRecord]:
+    fh.seek(offset)
+    header = _read_section_header(fh, c_ddf.Virtual_Disk_Records, c_ddf.DDF_VD_RECORD_SIGNATURE)
+    return [VirtualDiskRecord(fh) for _ in range(header.Populated_VDEs)]
+
+
+def _read_virtual_disk_configuration_records(
+    fh: BinaryIO, offset: int, count: int, size: int, num_entries: int
+) -> list[VirtualDiskConfigurationRecord]:
+    result = []
+    for _ in range(count):
+        fh.seek(offset)
+        obj = VirtualDiskConfigurationRecord(fh, num_entries)
+
+        if obj.header.Signature != c_ddf.DDF_VDCR_SIGNATURE:
+            continue
+
+        result.append(obj)
+        offset += size
+
+    return result
+
+
+def _convert_raid_layout(prl: int, rlq: int, srl: int, pec: int, sec: int) -> tuple[Level, Layout, int]:
+    """Convert DDF RAID layout to a layout we can work with.
+
+    Args:
+        prl: The DDF Primary RAID Level.
+        rlq: The DDF RAID Level Qualifier.
+        srl: The DDF Secondary RAID Level.
+        pec: The DDF Primary Element Count
+        sec: The DDF Secondary Element Count.
+    """
+    level = None
+    layout = None
+    num_disks = pec
+
+    if sec > 1:
+        if prl != c_ddf.DDF_VDCR_RAID1 or srl not in (c_ddf.DDF_VDCR_2STRIPED, c_ddf.DDF_VDCR_2SPANNED):
+            raise ValueError(f"Unsupported secondary RAID level: {srl}")
+
+        if num_disks == 2 and rlq == c_ddf.DDF_VDCR_RAID1_SIMPLE:
+            layout = 0x102  # 2 near copies, 1 far copy
+        elif num_disks == 3 and rlq == c_ddf.DDF_VDCR_RAID1_MULTI:
+            layout = 0x103  # 3 near copies, 1 far copy
+
+        num_disks *= sec
+        level = Level.RAID10
+
+    else:
+        if prl == c_ddf.DDF_VDCR_CONCAT:
+            level = Level.LINEAR
+
+        elif prl == c_ddf.DDF_VDCR_RAID0:
+            if rlq != c_ddf.DDF_VDCR_RAID0_SIMPLE:
+                raise ValueError(f"Unsupported DDF RAID layout: ({prl}, {rlq})")
```

Comment:
These error messages look all the same. I recommend including the RAID level in the string. At least then you know whether the RAID level has been detected correctly or in which code path you ended up otherwise (while debugging).